README.rst
xicorrelation
=============
.. image:: https://github.com/jettify/xicorrelation/workflows/CI/badge.svg
   :target: https://github.com/jettify/xicorrelation/actions?query=workflow%3ACI
   :alt: GitHub Actions status for master branch
.. image:: https://codecov.io/gh/jettify/xicorrelation/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/jettify/xicorrelation
.. image:: https://img.shields.io/pypi/pyversions/xicorrelation.svg
    :target: https://pypi.org/project/xicorrelation
.. image:: https://img.shields.io/pypi/v/xicorrelation.svg
    :target: https://pypi.python.org/pypi/xicorrelation
..
.. image:: https://readthedocs.org/projects/xicorrelation/badge/?version=latest
    :target: https://xicorrelation.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status


**xicorrelation** package implements xi correlation formula proposed in  https://arxiv.org/pdf/1909.10140.pdf.


It is based off the R code mentioned in the paper: https://statweb.stanford.edu/~souravc/xi.R and
R package https://github.com/cran/XICOR

+-----------------------------------------------------------------------------------------------+
| .. image:: https://raw.githubusercontent.com/jettify/xicorrelation/master/docs/anscombe.png   |
+-----------------------------------------------------------------------------------------------+


Simple example
--------------

.. code:: python

    from xicorrelation import xicorr

    x = [10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5]
    y = [8.04, 6.95, 7.58, 8.81, 8.33, 9.96, 7.24, 4.26, 10.84, 4.82, 5.68]

    # API similar to spearmanr or kendalltau from scipy
    xi, pvalue = xicorr(x, y)
    print("xi", xi)
    print("pvalue", pvalue)


Installation
------------
Installation process is simple, just::

    $ pip install xicorrelation


CHANGES.rst
Changes
-------

0.3.0
------------------
* Rework API do be similar to scipy.

0.2.1
------------------
* Initial release.


tests/test_basic.py
import random

import numpy as np
import pytest

from xicorrelation import xicorr


@pytest.fixture
def quartet():
    x_1 = [10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5]
    x_2 = [10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5]
    x_3 = [0, 0, 0, 0, 1, 1, 2, 3]
    x_4 = [8, 8, 8, 8, 8, 8, 8, 19, 8, 8, 8]
    y_1 = [8.04, 6.95, 7.58, 8.81, 8.33, 9.96, 7.24, 4.26, 10.84, 4.82, 5.68]
    y_2 = [9.14, 8.14, 8.74, 8.77, 9.26, 8.1, 6.13, 3.1, 9.13, 7.26, 4.74]
    y_3 = [0, 1, 2, 3, 4, 5, 6, 7]
    y_4 = [6.58, 5.76, 7.71, 8.84, 8.47, 7.04, 5.25, 12.5, 5.56, 7.91, 6.89]
    anscombes_quartet = {
        "x_1": x_1,
        "x_2": x_2,
        "x_3": x_3,
        "x_4": x_4,
        "y_1": y_1,
        "y_2": y_2,
        "y_3": y_3,
        "y_4": y_4,
    }

    return anscombes_quartet


def test_xi_correlations(quartet):
    q = quartet

    assert xicorr(q["x_1"], q["y_1"]).correlation == pytest.approx(
        0.275, rel=10 ** -6
    )
    assert xicorr(q["x_2"], q["y_2"]).correlation == pytest.approx(
        0.6, rel=10 ** -6
    )
    assert xicorr(q["x_3"], q["y_3"]).correlation == pytest.approx(
        0.6666666666666667, rel=10 ** -6
    )
    assert xicorr(q["x_4"], q["y_4"]).correlation == pytest.approx(
        0.175, rel=10 ** -6
    )


def test_xi_correlations_no_ties(quartet):
    q = quartet
    x = xicorr(q["x_1"], q["y_1"], ties=False)
    assert x.correlation == pytest.approx(0.275, rel=10 ** -6)


def test_p_val_asymptotic(quartet):
    random.seed(2020)
    np.random.seed(2020)
    q = quartet
    # values taken from R code
    assert xicorr(q["x_1"], q["y_1"], ties=True).pvalue == pytest.approx(
        0.0784155644
    )
    assert xicorr(q["x_2"], q["y_2"], ties=True).pvalue == pytest.approx(
        0.0010040217
    )

    assert xicorr(q["x_4"], q["y_4"], ties=True).pvalue == pytest.approx(
        0.1838021283
    )
    assert xicorr(q["x_3"], q["y_3"], ties=True).pvalue == pytest.approx(
        0.001986300
    )


def test_errors():
    with pytest.raises(ValueError):
        xicorr([1.0, 2.0, 3.0, 4.0], [5.0, 6.0])

    with pytest.raises(ValueError):
        xicorr(np.array([1, 2, 3, 4]), np.array([5, 6]))


def test_flatten():
    x = np.array([[1, 2], [3, 4]])
    y = np.array([[1, 4], [9, 16]])
    xi, pvalue = xicorr(x, y)
    assert xi == pytest.approx(0.4)
    assert pvalue == pytest.approx(0.13362874657719392)


def test_empty():
    xi, pvalue = xicorr([], [], ties=False)
    assert xi is np.nan
    assert pvalue is np.nan


examples/iris.py
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from scipy.stats import kendalltau, pearsonr, spearmanr
from sklearn import datasets

from xicorrelation import xicorr

iris = datasets.load_iris()
df = pd.DataFrame(iris.data)
df.columns = iris.feature_names

sns.set(style="ticks", color_codes=True)

g = sns.pairplot(df, palette="husl", markers=["o", "s", "D"])


def corrfunc(x, y, ax=None, **kw):
    pr = pearsonr(x, y)[0]
    sr = spearmanr(x, y).correlation
    kt = kendalltau(x, y).correlation
    xi = xicorr(x, y).correlation

    ax = ax or plt.gca()
    bbox = dict(boxstyle="round", fc="blanchedalmond", ec="orange", alpha=0.5)
    ax.annotate(
        f"Pearson r = {pr:.2f}",
        xy=(0.1, 0.9),
        xycoords=ax.transAxes,
        bbox=bbox,
        fontsize=7,
    )
    ax.annotate(
        f"Spearman r = {sr:.2f}",
        xy=(0.1, 0.8),
        xycoords=ax.transAxes,
        bbox=bbox,
        fontsize=7,
    )
    ax.annotate(
        f"Kendall tau = {kt:.2f}",
        xy=(0.1, 0.7),
        xycoords=ax.transAxes,
        bbox=bbox,
        fontsize=7,
    )
    ax.annotate(
        f"Chatterjee xi = {xi:.2f}",
        xy=(0.1, 0.6),
        xycoords=ax.transAxes,
        bbox=bbox,
        fontsize=7,
    )


g.map_lower(corrfunc)
g.map_upper(corrfunc)
plt.show()


xicorrelation/__init__.py
from typing import NamedTuple, Optional

import numpy as np
import numpy.typing as npt
from scipy.stats import norm, rankdata

from ._version import version as _version

__all__ = ("xicorr", "XiCorrResult")
__version__ = _version


class _XiCorr(NamedTuple):
    xi: float
    fr: npt.NDArray[np.float64]
    cu: float


class XiCorrResult(NamedTuple):
    correlation: float
    pvalue: Optional[float]


def _xicorr(x: npt.NDArray[np.float_], y: npt.NDArray[np.float_]) -> _XiCorr:
    # Ported from original R implementation.
    # https://github.com/cran/XICOR/blob/master/R/calculateXI.R
    n = x.size
    PI = rankdata(x, method="average")
    fr = rankdata(y, method="average") / n
    gr = rankdata(-y, method="average") / n
    cu = np.mean(gr * (1 - gr))
    A1 = np.abs(np.diff(fr[np.argsort(PI, kind="quicksort")])).sum() / (2 * n)
    xi = 1.0 - A1 / cu
    return _XiCorr(xi, fr, cu)


def xicorr(
    x: npt.ArrayLike, y: npt.ArrayLike, ties: bool = True
) -> XiCorrResult:
    """Compute the cross rank increment correlation coefficient xi [1].

    Parameters
    ----------
    x, y : array_like
        Arrays of rankings, of the same shape. If arrays are not 1-D, they
        will be flattened to 1-D.
    ties : bool, optional
        If ties is True, the algorithm assumes that the data has ties and
        employs the more elaborated theory for calculating s.d. and P-value.
        Otherwise, it uses the simpler theory. There is no harm in putting
        ties = True even if there are no ties.

    Returns
    -------
    correlation : float
       The tau statistic.
    pvalue : float
       P-values computed by the asymptotic theory.

    See Also
    --------
    spearmanr : Calculates a Spearman rank-order correlation coefficient.


    Example:
        >>> from xicorrelation import xicorr
        >>> x = [1, 2, 3, 4, 5]
        >>> y = [1, 4, 9, 16, 25]
        >>> xi, pvalue = xicorr(x, y)
        >>> print(xi, pvalue)

    References
    ----------
    .. [1] Chatterjee, S., "A new coefficient of correlation",
           https://arxiv.org/abs/1909.10140, 2020.

    Examples
    --------
    >>> from scipy import stats
    >>> x1 = [12, 2, 1, 12, 2]
    >>> x2 = [1, 4, 7, 1, 0]
    >>> xi, p_value, _ = xicorr(x1, x2)
    >>> tau
    -0.47140452079103173
    >>> p_value
    0.2827454599327748
    """
    # https://git.io/JSIlN
    x = np.asarray(x).ravel()
    y = np.asarray(y).ravel()

    if x.size != y.size:
        raise ValueError(
            "All inputs to `xicorr` must be of the same "
            f"size, found x-size {x.size} and y-size {y.size}"
        )
    elif not x.size or not y.size:
        # Return NaN if arrays are empty
        return XiCorrResult(np.nan, np.nan)

    r = _xicorr(x, y)
    xi = r.xi
    fr = r.fr
    CU = r.cu

    pvalue = None
    # https://git.io/JSIlM
    n = x.size
    if not ties:
        # sd = np.sqrt(2.0 / (5.0 * n))
        pvalue = 1.0 - norm.cdf(np.sqrt(n) * xi / np.sqrt(2.0 / 5.0))
    else:
        qfr = np.sort(fr)
        ind = np.arange(1, n + 1)
        ind2 = 2 * n - 2 * ind + 1

        ai = np.mean(ind2 * qfr * qfr) / n
        ci = np.mean(ind2 * qfr) / n
        cq = np.cumsum(qfr)
        m = (cq + (n - ind) * qfr) / n
        b = np.mean(m ** 2)
        v = (ai - 2.0 * b + ci ** 2) / (CU ** 2)

        # sd = np.sqrt(v / n)
        pvalue = 1.0 - norm.cdf(np.sqrt(n) * xi / np.sqrt(v))
    return XiCorrResult(xi, pvalue)


examples/anscombe.py
import matplotlib.pyplot as plt
import numpy as np
from scipy.stats import kendalltau, pearsonr, spearmanr

from xicorrelation import xicorr

x = [10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5]
y1 = [8.04, 6.95, 7.58, 8.81, 8.33, 9.96, 7.24, 4.26, 10.84, 4.82, 5.68]
y2 = [9.14, 8.14, 8.74, 8.77, 9.26, 8.10, 6.13, 3.10, 9.13, 7.26, 4.74]
y3 = [7.46, 6.77, 12.74, 7.11, 7.81, 8.84, 6.08, 5.39, 8.15, 6.42, 5.73]
x4 = [8, 8, 8, 8, 8, 8, 8, 19, 8, 8, 8]
y4 = [6.58, 5.76, 7.71, 8.84, 8.47, 7.04, 5.25, 12.50, 5.56, 7.91, 6.89]

datasets = {"I": (x, y1), "II": (x, y2), "III": (x, y3), "IV": (x4, y4)}

fig, axs = plt.subplots(
    2,
    2,
    sharex=True,
    sharey=True,
    figsize=(6, 6),
    gridspec_kw={"wspace": 0.08, "hspace": 0.08},
)
axs[0, 0].set(xlim=(0, 20), ylim=(2, 14))
axs[0, 0].set(xticks=(0, 10, 20), yticks=(4, 8, 12))

for ax, (label, (x, y)) in zip(axs.flat, datasets.items()):
    ax.text(0.1, 0.9, label, fontsize=20, transform=ax.transAxes, va="top")
    ax.tick_params(direction="in", top=True, right=True)
    ax.plot(x, y, "o")

    # linear regression
    p1, p0 = np.polyfit(x, y, deg=1)  # slope, intercept
    ax.axline(xy1=(0, p0), slope=p1, color="r", lw=2)

    pr = pearsonr(x, y)[0]
    sr = spearmanr(x, y).correlation
    kt = kendalltau(x, y).correlation
    xi = xicorr(x, y).correlation

    # add text box for the statistics
    stats = (
        f"$\\mu$ = {np.mean(y):.2f}\n"
        f"$\\sigma$ = {np.std(y):.2f}\n"
        # correlations
        f"Pearson $r$ = {pr:.4f}\n"
        f"Spearmanr $r$ = {pr:.4f}\n"
        f"Kendall $tau$ = {kt:.4f}\n"
        f"Chatterjee $xi$ = {xi:.4f}"
    )
    bbox = dict(boxstyle="round", fc="blanchedalmond", ec="orange", alpha=0.0)
    ax.text(
        0.95,
        0.07,
        stats,
        fontsize=8,
        bbox=bbox,
        transform=ax.transAxes,
        horizontalalignment="right",
    )
fig.suptitle("Anscombe Quadrant")
plt.show()
